### ISSUE TITLE
(Linux Mint) VPN's DNS server is missing after waking the system from suspend/sleep state

### ISSUE THREAD

**Issue by edwardivpn** [ivpn/desktop-app#195]
# Bug report

## Describe your environment

* OS name and version: Linux Mint 20.3
* IVPN app version: 3.9.0 

## Describe the problem

After waking the device from suspend/sleep state, the system is missing the applied previously IVPN's DNS server.  This results in hostname resolution failure (with IVPN Firewall enabled) or may expose the system to potential leaks (with Firewall disabled).

The IVPN's DNS is reapplied by reconnecting to the VPN server.

Reportedly, the issue has started occurring ~2 weeks ago.

I have managed to replicate the issue on my Mint 20.3 VM. The issue is not reproducible on Ubuntu, or Fedora.

I have not tested the previous version of our app. The issue occurs regardless of the used VPN protocol.

Before the system goes into a sleep/suspend state:

```
ivpn status

VPN : CONNECTED
ch.wg.ivpn.net, Zurich (CH), Switzerland
Protocol : WireGuard
Local IP : 172.xx.xx.xx
Local IPv6 : fd00:4956:504e:ffff::xx:xx
Server IP : 185.212.170.141
Connected : 2022-08-03 14:55:54 +0200 CEST
AntiTracker : Enabled
Split Tunnel : Enabled
Firewall : Enabled
Allow LAN : false
Persistent : true
Allow IVPN servers : true

---

ping 1.1.1.1

PING 1.1.1.1 (1.1.1.1) 56(84) bytes of data.
64 bytes from 1.1.1.1: icmp_seq=3D2 ttl=3D59 time=3D55.1 ms
64 bytes from 1.1.1.1: icmp_seq=3D3 ttl=3D59 time=3D161 ms
64 bytes from 1.1.1.1: icmp_seq=3D4 ttl=3D59 time=3D124 ms
64 bytes from 1.1.1.1: icmp_seq=3D5 ttl=3D59 time=3D36.8 ms

---

cat /run/systemd/resolve/resolv.conf

nameserver 192.168.1.1
nameserver 10.0.254.2

---

systemctl status systemd-resolved.service

● systemd-resolved.service - Network Name Resolution
      Loaded: loaded (/lib/systemd/system/systemd-resolved.service; 
enabled; vendor preset: enabled)
      Active: active (running) since Thu 2022-08-04 11:13:39 CEST; 8min ago
        Docs: man:systemd-resolved.service(8)
https://www.freedesktop.org/wiki/Software/systemd/resolved
https://www.freedesktop.org/wiki/Software/systemd/writing-network-configuration-managers
https://www.freedesktop.org/wiki/Software/systemd/writing-resolver-clients
    Main PID: 820 (systemd-resolve)
      Status: "Processing requests..."
       Tasks: 1 (limit: 9103)
      Memory: 9.0M
      CGroup: /system.slice/systemd-resolved.service
              └─820 /lib/systemd/systemd-resolved

Aug 04 11:13:39 ***-PC systemd[1]: Starting Network Name Resolution...
Aug 04 11:13:39 ***-PC systemd-resolved[820]: Positive Trust Anchors:
Aug 04 11:13:39***-PC systemd-resolved[820]: . IN DS 20326 82e06d44b80b8f1d39a95c0b0d7c65d08458e880409bbc683457104237c7f8ec8d
Aug 04 11:13:39 ***-PC systemd-resolved[820]: Negative trust anchors: 
10.in-addr.arpa 16.172.in-addr.arpa 17.172.in-addr.arpa 
18.172.in-addr.arpa 19.172.in-addr.arpa 20.172.in-addr.arpa 
21.172.in-addr.arpa 22.172.in-addr.arpa 23 24.172.in-addr.arpa 
25.172.in-addr.arpa 26.172.in-addr.arpa 27.172.in-addr.arpa 
28.172.in-addr.arpa 29.172.in-addr.arpa 30.172.in-addr.arpa 
31.172.in-addr.arpa 168.192.in-addr.arpa d.f.ip6.arpa corp home internal 
intranet lan local private test
Aug 04 11:13:39 ***-PC systemd-resolved[820]: Using system 
hostname '***-PC'.
Aug 04 11:13:39 ***-PC systemd[1]: Started Network Name Resolution.

---

systemd-resolve --status

Global
        LLMNR setting: no
MulticastDNS setting: no
   DNSOverTLS setting: no
       DNSSEC setting: no
     DNSSEC supported: no
           DNSSEC NTA: 10.in-addr.arpa
                       16.172.in-addr.arpa
                       168.192.in-addr.arpa
                       17.172.in-addr.arpa
                       18.172.in-addr.arpa
                       19.172.in-addr.arpa
                       20.172.in-addr.arpa
                       21.172.in-addr.arpa
                       22.172.in-addr.arpa
                       23.172.in-addr.arpa
                       24.172.in-addr.arpa
                       25.172.in-addr.arpa
                       26.172.in-addr.arpa
                       27.172.in-addr.arpa
                       28.172.in-addr.arpa
                       29.172.in-addr.arpa
                       30.172.in-addr.arpa
                       31.172.in-addr.arpa
                       corp
                       d.f.ip6.arpa
                       home
                       internal
                       intranet
                       lan
                       local
                       private
                       test

Link 4 (wgivpn)
       Current Scopes: DNS
DefaultRoute setting: yes
        LLMNR setting: yes
MulticastDNS setting: no
   DNSOverTLS setting: no
       DNSSEC setting: no
     DNSSEC supported: no
   Current DNS Server: 10.0.254.2
          DNS Servers: 10.0.254.2
           DNS Domain: ~.

Link 2 (wlp107s0)
       Current Scopes: DNS
DefaultRoute setting: yes
        LLMNR setting: yes
MulticastDNS setting: no
   DNSOverTLS setting: no
       DNSSEC setting: no
     DNSSEC supported: no
   Current DNS Server: 192.168.1.1
          DNS Servers: 192.168.1.1
             DNS Domain: ~.
```

After waking the device from sleep/suspend state:

```
ivpn status

VPN : CONNECTED
ch.wg.ivpn.net, Zurich (CH), Switzerland
Protocol : WireGuard
Local IP : 172.xx.xx.xx
Local IPv6 : fd00:4956:504e:ffff::xx:xx
Server IP : 185.212.170.141
Connected : 2022-08-03 14:55:54 +0200 CEST
AntiTracker : Enabled
Split Tunnel : Enabled
Firewall : Enabled
Allow LAN : false
Persistent : true
Allow IVPN servers : true

---

nslookup ivpn.net

;; connection timed out; no servers could be reached

---

ping 1.1.1.1

PING 1.1.1.1 (1.1.1.1) 56(84) bytes of data.
64 bytes from 1.1.1.1: icmp_seq=3D2 ttl=3D59 time=3D55.1 ms
64 bytes from 1.1.1.1: icmp_seq=3D3 ttl=3D59 time=3D161 ms
64 bytes from 1.1.1.1: icmp_seq=3D4 ttl=3D59 time=3D124 ms
64 bytes from 1.1.1.1: icmp_seq=3D5 ttl=3D59 time=3D36.8 ms
64 bytes from 1.1.1.1: icmp_seq=3D7 ttl=3D59 time=3D45.9 ms
^C
--- 1.1.1.1 ping statistics ---
7 packets transmitted, 5 received, 28.5714% packet loss, time 6059ms
rtt min/avg/max/mdev =3D 36.761/84.467/160.935/48.986 ms

---

cat /run/systemd/resolve/resolv.conf

nameserver 192.168.1.1

---

systemctl status systemd-resolved.service

systemd-resolved.service - 
Network Name Resolution Loaded: loaded (/lib/systemd/system/systemd-resolved.service; enabled; vendor preset: 
enabled) 
Active: active (running) since Wed 2022-08-03 12:54:19 CEST; 2h 44min ago 
Docs: man:systemd-resolved.service(8) 
https://www.freedesktop.org/wiki/Software/systemd/resolved 
https://www.freedesktop.org/wiki/Software/systemd/writing-network-configuration-managers 
https://www.freedesktop.org/wiki/Software/systemd/writing-resolver-clients 
Main PID: 823 (systemd-resolve) Status: "Processing requests..." Tasks: 
1 (limit: 9103) Memory: 9.1M CGroup: 
/system.slice/systemd-resolved.service └─823 
/lib/systemd/systemd-resolved 
Aug 03 12:54:19 XXX-PC systemd[1]: Starting Network Name Resolution... 
Aug 03 12:54:19 XXX-PC systemd-resolved[823]: Positive Trust Anchors: 
Aug 03 12:54:19 XXX-PC systemd-resolved[823]: . IN DS 20326 82e06d44b80b8f1d39a95c0b0d7c65d08458e880409bbc683457104237c7f8ec8d 
Aug 03 12:54:19 XXX-PC systemd-resolved[823]: Negative trust anchors: 
10.in-addr.arpa 16.172.in-addr.arpa 17.172.in-addr.arpa 
18.172.in-addr.arpa 19.172.in-addr.arpa 20.172.in-addr.arpa 
21.172.in-addr.arpa 22.172.in-addr.arpa 23.172.in-addr.arpa 
24.172.in-addr.arpa 25.172.in-addr.arpa 26.172.in-addr.arpa 
27.172.in-addr.arpa 28.172.in-addr.arpa 29.172.in-addr.arpa 
30.172.in-addr.arpa 31.172.in-addr.arpa 168.192.in-addr.arpa
d.f.ip6.arpa corp home internal intranet lan local private test
Aug 03 12:54:19 XXX-PC systemd-resolved[823]: Using system hostname 'XXX-PC'. 
Aug 03 12:54:19 XXX-PC systemd[1]: Started Network Name Resolution. 
Aug 03 12:54:36 XXX-PC systemd-resolved[823]: Using degraded feature set (UDP) for DNS server 10.0.254.2. 
Aug 03 13:15:20 XXX-PC systemd-resolved[823]: Grace period over, resuming full feature set (UDP+EDNS0) for DNS server 10.0.254.2.

---

Global
       LLMNR setting: no                 
MulticastDNS setting: no                 
  DNSOverTLS setting: no                 
      DNSSEC setting: no                 
    DNSSEC supported: no                 
          DNSSEC NTA: 10.in-addr.arpa    
                      16.172.in-addr.arpa
                      168.192.in-addr.arpa
                      17.172.in-addr.arpa
                      18.172.in-addr.arpa
                      19.172.in-addr.arpa
                      20.172.in-addr.arpa
                      21.172.in-addr.arpa
                      22.172.in-addr.arpa
                      23.172.in-addr.arpa
                      24.172.in-addr.arpa
                      25.172.in-addr.arpa
                      26.172.in-addr.arpa
                      27.172.in-addr.arpa
                      28.172.in-addr.arpa
                      29.172.in-addr.arpa
                      30.172.in-addr.arpa
                      31.172.in-addr.arpa
                      corp               
                      d.f.ip6.arpa       
                      home               
                      internal           
                      intranet           
                      lan                
                      local              
                      private            
                      test               

Link 5 (wgivpn)
      Current Scopes: none
DefaultRoute setting: yes
       LLMNR setting: yes
MulticastDNS setting: no 
  DNSOverTLS setting: no 
      DNSSEC setting: no 
    DNSSEC supported: no 

Link 2 (wlp107s0)
      Current Scopes: DNS       
DefaultRoute setting: yes       
       LLMNR setting: yes       
MulticastDNS setting: no        
  DNSOverTLS setting: no       
      DNSSEC setting: no        
    DNSSEC supported: no        
  Current DNS Server: 192.168.1.1
         DNS Servers: 192.168.1.1
          DNS Domain: ~.        

```

### Steps to reproduce:

1. Connect to any VPN server
2. Suspend/Sleep the system, e.g. with `systemctl suspend` command
3. Wake the system from the suspend/sleep state

### Observed Results:

The previously applied IVPN DNS is missing causing the hostname resolution to fail

### Expected Results:

IVPN DNS server remains applied through the suspend/sleep state

**Comment by stenya** [comment]
The issue can be also reproduced by simply disable/enable connectivity in OS settings (in my case it was disable/enable Wired connection on Linux Mint 20; proxmox VM#124 )

**Comment by stenya** [comment]
v3.9.6

**Comment by gorkapernas** [comment]
Reproduced the issue on 3.9.0 and I confirmed that the issue is fixed on version 3.9.6, tested on Linux Mint 20 Cinnamon.
VPN is connected and DNS remains applied after awaking the machine or enabling back the Wired connection.
This is good to go.

**Comment by stenya** [comment]
Released in v3.9.8
